### PR TITLE
Prepare ComInterfaceGenerator for going into the public API

### PR DESCRIPTION
**Pull request by jkoritzinsky**
- Rename the exception marshallers per API review.
- Since we aren't going to expose the "object unwrapper" mechanism until we expose the generic "vtable" logic, change the ComInterfaceGenerator to use the ComWrappers.ComInterfaceDispatch APIs directly.
- Separate more of the code between the VirtualMethodIndex generator and the ComInterface generator so we don't have the unsupported generator code influencing the shipping generator's code.

### CODE REVIEW

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Marshallers/ComInterfaceDispatchMarshallerFactory.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+        public IMarshallingGenerator Create(TypePositionInfo info, StubCodeContext context)
+            => info.MarshallingAttributeInfo is ComInterfaceDispatchMarshallingInfo ? new Marshaller() : _inner.Create(info, context);
+
+        private sealed class Marshaller : IMarshallingGenerator
+        {
+            public ManagedTypeInfo AsNativeType(TypePositionInfo info) =>
+                new PointerTypeInfo(
+                    $"{TypeNames.System_Runtime_InteropServices_ComWrappers_ComInterfaceDispatch}*",
+                    $"{TypeNames.System_Runtime_InteropServices_ComWrappers_ComInterfaceDispatch}*",
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Marshallers/ComInterfaceDispatchMarshallerFactory.cs:34]: I can't recall, but are these supposed to be the same? I thought one was "to" and the other was "from".

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Marshallers/ComInterfaceDispatchMarshallerFactory.cs [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+                                    TypeArgumentList(SingletonSeparatedList(info.ManagedType.Syntax)))),
+                            ArgumentList(
+                                SingletonSeparatedList(
+                                    Argument(
+                                        IdentifierName(native)))))));
+            }
+
+            public SignatureBehavior GetNativeSignatureBehavior(TypePositionInfo info) => SignatureBehavior.NativeType;
+            public ValueBoundaryBehavior GetValueBoundaryBehavior(TypePositionInfo info, StubCodeContext context) => ValueBoundaryBehavior.NativeIdentifier;
+            public bool IsSupported(TargetFramework target, Version version) => true;
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Marshallers/ComInterfaceDispatchMarshallerFactory.cs:63]: Seems like this needs to be limited to .NET 8, right?

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Marshallers/ComInterfaceDispatchMarshallerFactory.cs (hunk 1 above)
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Marshallers/ComInterfaceDispatchMarshallerFactory.cs:34]: The names here are the "full name" and the "diagnostic full name" for reporting in diagnostics to the user. They're the same in most cases.

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Marshallers/ComInterfaceDispatchMarshallerFactory.cs (hunk 2 above)
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/Marshallers/ComInterfaceDispatchMarshallerFactory.cs:63]: We can limit this to .NET 5 or newer as that's the first version where the API we're using here was introduced.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Pull request merged**